src/core/time.jl

```diff
@@ -72,150 +108,318 @@ const AnyIndex = Union{Int, Vector{Int}, Tuple, Colon, OrdinalRange}
 end
 
+# TBD:  this function assumes fixed step size, need to parameterize properly 
+# and then create a version for variable timestep.  It is also not used within
+# the code and possibly incorrectly interprets the meaning of the Base.indices function.
+# function Base.indices(x::TimestepVector{T, Start, Step}) where {T, Start, Step}
+# 	return (Start:Step:(Start + (length(x.data) - 1) * Step), )
+# end
+
 function Base.length(v::TimestepVector)
 	return length(v.data)
 end
 
 Base.endof(v::TimestepVector) = length(v)
 
 #
-# TimestepMatrix
+# 3c. TimestepMatrix
 #
-function Base.getindex(mat::TimestepMatrix{T, Start, Step}, ts::Timestep{Start, Step, Stop}, i::AnyIndex) where {T, Start, Step, Stop}
+
+
+# function Base.getindex(mat::TimestepMatrix, ts::T, i::AnyIndex) where {T <: AbstractTimestep}
+# 	return mat.data[ts.t, i]
+# end
+
+function Base.getindex(mat::TimestepMatrix{Timestep{Start, Step}, T}, ts::Timestep{Start, Step, Stop}, i::AnyIndex) where {T, Start, Step, Stop} 
+	return mat.data[ts.t, i]
+end
+
+function Base.getindex(mat::TimestepMatrix{VariableTimestep{start_times}, T}, ts::VariableTimestep{start_times}, i::AnyIndex) where {T, start_times}
 	return mat.data[ts.t, i]
 end
 
-function Base.getindex(mat::TimestepMatrix{T, d_start, Step}, ts::Timestep{t_start, Step, Stop}, i::AnyIndex) where {T, d_start, Step, t_start, Stop}
+function Base.getindex(mat::TimestepMatrix{Timestep{d_start, Step}, T}, ts::Timestep{t_start, Step, Stop}, i::AnyIndex) where {T, d_start, t_start, Step, Stop} 
 	t = Int(ts.t + (t_start - d_start) / Step)
-	return mat.data[t, i]
+	return return mat.data[t, i]
+end
+
+function Base.getindex(mat::TimestepMatrix{VariableTimestep{d_start_times}, T}, ts::VariableTimestep{t_start_times}, i::AnyIndex) where {T, d_start_times, t_start_times}
+	t = ts.t + findfirst(d_start_times, t_start_times[1]) - 1
+	return return mat.data[t, i]
 end
 
 # int indexing version supports old style components
-function Base.getindex(mat::TimestepMatrix{T, Start, Step}, idx1::AnyIndex, idx2::AnyIndex) where {T, Start, Step}
+# function Base.getindex(mat::TimestepMatrix, idx1::AnyIndex, idx2::AnyIndex)
+# 	return mat.data[idx1, idx2]
+# end
+
+function Base.getindex(mat::TimestepMatrix{Timestep{Start, Step}, T}, idx1::AnyIndex, idx2::AnyIndex) where {T, Start, Step}
+	return mat.data[idx1, idx2]
+end
+
+function Base.getindex(mat::TimestepMatrix{VariableTimestep{start_times}, T}, idx1::AnyIndex, idx2::AnyIndex) where {T, start_times}
 	return mat.data[idx1, idx2]
 end
 
-function Base.setindex!(mat::TimestepMatrix{T, Start, Step}, val, ts::Timestep{Start, Step, Stop}, idx::AnyIndex) where {T, Start, Step, Stop}
+# function Base.setindex(mat::TimestepMatrix, val, ts::T, idx::AnyIndex) where {T <: AbstractTimestep}
+# 	setindex!(mat.data, val, ts.t, idx)
+# end
+
+function Base.setindex(mat::TimestepMatrix{Timestep{Start, Step}, T}, val, ts::Timestep{Start, Step, Stop}, idx::AnyIndex) where {T, Start, Step, Stop} 
```

Comment:
Just had a thought (OH NO): normally we use all caps for type parameters... Wondering whether we should do the same throughout, i.e. ``START``, ``STEP`` and ``STOP`` etc...

Maybe we can open an issue for that, we shouldn't stop this PR from merging for that, but we might want to change it down the road.